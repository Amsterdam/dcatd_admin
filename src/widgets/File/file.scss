@import "../../styling/config.scss";

.file {
  &__upload {
    width: 60% !important;

    &-label {
      background-color: $darkish-blue;
      color: $white;
      display: inline-block;
      padding: 4px 6px;
    }

    &--hidden {
      opacity: 0;
      position: absolute;
    }
  }

  &__progress {
    border: 1px  solid $warm-grey;
    position: relative;
    width: 60%;

    &-percentage {
      background-color: $white-two;
      height: 38px;
    }

    &-filename {
      left: 16px;
      position: absolute;
      top: 9px;
    }

    &-close-button {
      @extend %icon-verwijder;
      background-color: transparent;
      border-style: none;
      height: 30px;
      position: absolute;
      right: 4px;
      top: 4px;
      width: 30px;

      &:hover {
        @extend %icon-verwijder-white;
        background-color: $red;
      }
    }
  }

  &__purl {
    background-color: $white-four;
    height: auto;
    margin-top: 4px;
    overflow: auto;
    padding: 4px 10px;

    &-title {
      @extend %heavy-font;

      margin-top: 16px;
    }
  }

  .form-control {
    margin-bottom: 10px;
  }
}
