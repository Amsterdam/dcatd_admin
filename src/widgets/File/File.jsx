import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import defer from 'lodash.defer';

import api from '../../services/api/api';
import serverError from '../../services/server-error/server-error';
import { getAccessToken } from '../../services/auth/auth';

import './file.scss';

const mapDispatchToProps = dispatch => bindActionCreators({
  serverError
}, dispatch);

class File extends Component {
  constructor(props) {
    super(props);
    this.state = {
      file: props.file,
      loaded: props.loaded,
      status: props.status,
      total: props.total,
      purl: props.purl,
      url: props.url,
      value: props.value
    };

    this.processFile = this.processFile.bind(this);
    this.resetFile = this.resetFile.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  componentWillReceiveProps(props) {
    this.setState({
      purl: props.purl,
      value: props.value
    });

    if (!props.value) {
      this.resetFile();
    }
  }

  processFile(files) {
    if (files.length !== 1) {
      return;
    }

    const formData = new window.FormData();
    formData.append('distribution', files[0]);

    const xhr = new window.XMLHttpRequest();

    xhr.onloadstart = () => {
      this.setState({
        status: 'busy',
        file: files[0].name
      });

      if (this.props.registry.formContext && this.props.registry.formContext.setUploadStatus) {
        this.props.registry.formContext.setUploadStatus('busy');
      }
    };

    xhr.upload.onprogress = (e) => {
      this.setState({
        loaded: e.loaded,
        total: e.total
      });
    };

    xhr.onload = () => {
      // upload success
      if (xhr.readyState === 4 && xhr.status === 201 && xhr.statusText === 'Created') {
        const location = xhr.getResponseHeader('Location');
        this.setState({
          status: 'done',
          url: location,
          value: location
        });

        if (this.props.registry.formContext && this.props.registry.formContext.setResourceSpecs) {
          this.props.registry.formContext.setResourceSpecs({
            'dcat:byteSize': this.state.total,
            'dct:format': files[0].type,
            'ams:distributionType': 'file'
          });
        }

        this.props.onChange(location);

        defer(() => {
          this.handleChange(location);
        });
      } else {
        this.props.serverError(xhr);

        this.resetFile();
      }
    };

    xhr.open('POST', api.files, true);
    xhr.setRequestHeader('Authorization', `Bearer ${getAccessToken()}`);
    xhr.send(formData);
  }

  resetFile() {
    this.setState({
      file: '',
      loaded: 0,
      total: 0,
      status: 'idle',
      url: ''
    });

    if (this.props.registry.formContext && this.props.registry.formContext.setUploadStatus) {
      this.props.registry.formContext.setUploadStatus('idle');
    }
  }

  calculateProgress() {
    if (this.state.status === 'done') {
      return 100;
    }
    return (this.state.total ? ((this.state.loaded / this.state.total) * 95).toFixed(0) : 0);
  }

  handleChange(value) {
    this.setState({
      value
    });
    this.props.onChange(value || undefined);
  }

  render() {
    const { file, purl, status, value } = this.state;
    return (
      <div className="file">
        <input
          type="text"
          id={this.props.id}
          className="form-control file__url"
          label={this.props.label}
          disabled={this.props.disabled}
          placeholder={this.props.placeholder}
          readOnly={this.props.readonly}
          required={this.props.required}
          value={value}
          onChange={event => this.handleChange(event.target.value)}
        />
        <label
          htmlFor={`${this.props.id}-upload`}
          className={`
            file__upload-label
            ${status === 'idle' ? '' : 'file__upload--hidden'}
          `}
        >Selecteer bestand voor upload</label>
        <input
          type="file"
          className="file__upload file__upload--hidden"
          id={`${this.props.id}-upload`}
          disabled={this.props.disabled}

          onChange={event => this.processFile(event.target.files)}
        />
        {status !== 'idle' &&
          <div className="file__progress">
            <div
              className="file__progress-percentage"
              style={{ width: `${this.calculateProgress()}%` }}
            />
            <span className="file__progress-filename">{file}</span>
            {status === 'done' ?
              <button
                className="file__progress-close-button"
                onClick={() => this.resetFile()}
              /> : ''}
          </div>
        }
        {purl &&
          <div>
            <div className="file__purl-title">Permanente URL</div>
            <div className="file__purl">
              <p>{purl}</p>
            </div>
          </div>
        }
      </div>
    );
  }
}

File.defaultProps = {
  disabled: false,
  file: '',
  label: '',
  placeholder: '',
  loaded: 0,
  readonly: false,
  registry: {},
  required: false,
  status: 'idle',
  total: 0,
  purl: '',
  url: '',
  value: '',

  onChange: () => {},
  serverError: () => {}
};

File.propTypes = {
  disabled: PropTypes.bool,
  file: PropTypes.string,
  id: PropTypes.string.isRequired,
  label: PropTypes.string,
  loaded: PropTypes.number,
  placeholder: PropTypes.string,
  readonly: PropTypes.bool,
  registry: PropTypes.object,
  required: PropTypes.bool,
  status: PropTypes.string,
  total: PropTypes.number,
  purl: PropTypes.string,
  url: PropTypes.string,
  value: PropTypes.string,

  onChange: PropTypes.func,
  serverError: PropTypes.func
};

export default connect(
  null,
  mapDispatchToProps
)(File);
