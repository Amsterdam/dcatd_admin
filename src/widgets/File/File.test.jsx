import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import File from './File';
import { shallow } from '../../../test/enzyme';

const middlewares = [thunk];

jest.mock('../../services/auth/auth.js');

describe('The File component (react-jsonschema-forms custom field)', () => {
  let oldXMLHttpRequest;
  let oldFormData;
  let store;
  let wrap;
  let xhrSpy;
  let appendSpy;

  beforeEach(() => {
    xhrSpy = {
      open: jest.fn(),
      setRequestHeader: jest.fn(),
      send: jest.fn(),
      onloadstart: jest.fn(),
      onload: jest.fn(),
      onprogress: jest.fn()
    };
    const xhrMock = () => ({
      open: xhrSpy.open,
      setRequestHeader: xhrSpy.setRequestHeader,
      send: xhrSpy.send,
      onloadstart: xhrSpy.onloadstart,
      onload: xhrSpy.onload,
      upload: { onprogress: xhrSpy.onprogress }
    });
    oldXMLHttpRequest = window.XMLHttpRequest;
    window.XMLHttpRequest = jest.fn().mockImplementation(xhrMock);

    appendSpy = jest.fn();
    const formDataMock = () => ({
      append: appendSpy
    });
    oldFormData = window.FormData;
    window.FormData = jest.fn().mockImplementation(formDataMock);

    const props = {
      id: 'very-unique-id',
      registry: {
        formContext: {
          setResourceSpecs: jest.fn(),
          setUploadStatus: jest.fn()
        }
      }
    };

    store = configureMockStore(middlewares)();
    wrap = shallow(<File {...props} />, { context: { store } }).dive();
  });

  afterEach(() => {
    window.XMLHttpRequest = oldXMLHttpRequest;
    window.FormData = oldFormData;
  });

  // describe('renders component', () => {
  //   it('renders default', () => {
  //     expect(wrap).toMatchSnapshot();
  //   });
  // });

  describe('triggering events', () => {
    it('start uploadig a file when a new file is added to file input', () => {
      const event = { target: { files: ['upload-file.pdf'] } };

      wrap.find('#very-unique-id-upload').props().onChange(event);

      expect(xhrSpy.open).toHaveBeenCalledWith('POST', 'https://acc.api.data.amsterdam.nl/dcatd/files', true);
      expect(xhrSpy.setRequestHeader).toHaveBeenCalled();
      expect(xhrSpy.send).toHaveBeenCalledWith({ append: expect.any(Function) });

      expect(appendSpy).toHaveBeenCalledWith('distribution', 'upload-file.pdf');
    });
  });

  it('will not start uploadig a file when a no file is added to file input', () => {
    const event = { target: { files: [] } };

    wrap.find('#very-unique-id-upload').props().onChange(event);

    expect(xhrSpy.open).not.toHaveBeenCalled();
    expect(xhrSpy.setRequestHeader).not.toHaveBeenCalled();
    expect(xhrSpy.send).not.toHaveBeenCalled();

    expect(appendSpy).not.toHaveBeenCalled();
  });
});
