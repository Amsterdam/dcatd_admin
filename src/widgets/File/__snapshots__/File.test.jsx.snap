// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The File component (react-jsonschema-forms custom field) renders component renders default 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <File
    disabled={false}
    file=""
    id="very-unique-id"
    label=""
    loaded={0}
    onChange={[Function]}
    placeholder=""
    purl=""
    readonly={false}
    registry={Object {}}
    required={false}
    serverError={[Function]}
    status="idle"
    total={0}
    url=""
    value=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="form-control file__url"
          disabled={false}
          id="very-unique-id"
          label=""
          onChange={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
          type="text"
          value=""
        />,
        <label
          className="
            file__upload-label
            
          "
          htmlFor="very-unique-id-upload"
        >
          Selecteer bestand voor upload
        </label>,
        <input
          className="file__upload file__upload--hidden"
          disabled={false}
          id="very-unique-id-upload"
          onChange={[Function]}
          type="file"
        />,
        false,
        "",
      ],
      "className": "file",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "form-control file__url",
          "disabled": false,
          "id": "very-unique-id",
          "label": "",
          "onChange": [Function],
          "placeholder": "",
          "readOnly": false,
          "required": false,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Selecteer bestand voor upload",
          "className": "
            file__upload-label
            
          ",
          "htmlFor": "very-unique-id-upload",
        },
        "ref": null,
        "rendered": "Selecteer bestand voor upload",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "file__upload file__upload--hidden",
          "disabled": false,
          "id": "very-unique-id-upload",
          "onChange": [Function],
          "type": "file",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      false,
      "",
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="form-control file__url"
            disabled={false}
            id="very-unique-id"
            label=""
            onChange={[Function]}
            placeholder=""
            readOnly={false}
            required={false}
            type="text"
            value=""
          />,
          <label
            className="
            file__upload-label
            
          "
            htmlFor="very-unique-id-upload"
          >
            Selecteer bestand voor upload
          </label>,
          <input
            className="file__upload file__upload--hidden"
            disabled={false}
            id="very-unique-id-upload"
            onChange={[Function]}
            type="file"
          />,
          false,
          "",
        ],
        "className": "file",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "form-control file__url",
            "disabled": false,
            "id": "very-unique-id",
            "label": "",
            "onChange": [Function],
            "placeholder": "",
            "readOnly": false,
            "required": false,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Selecteer bestand voor upload",
            "className": "
            file__upload-label
            
          ",
            "htmlFor": "very-unique-id-upload",
          },
          "ref": null,
          "rendered": "Selecteer bestand voor upload",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "file__upload file__upload--hidden",
            "disabled": false,
            "id": "very-unique-id-upload",
            "onChange": [Function],
            "type": "file",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        false,
        "",
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
